### PR TITLE
evaluator multiple inheritance

### PR DESCRIPTION
**Pull request by nicolewhite**

### CODE REVIEW

File: test/testing/run.spec.ts [hunk 1]
```diff
@@ -843,6 +849,106 @@ describe('Testing SDK', () => {
...
   });
 
+  it('handles evaluators that implement both BaseTestEvaluator and BaseEventEvaluator', async () => {
+    type T = { x: number };
+    type O = string;
+
+    class MyEvaluator
+      extends BaseTestEvaluator<T, O>
+      implements BaseEventEvaluator
+    {
```
Comment (nicolewhite) [inline test/testing/run.spec.ts:859]: i think what we can do here is provide a `BaseEvaluator` ABC that combines both so you don't need to do this "extends" and then "implements" workaround

File: src/testing/models.ts [hunk 2]
```diff
@@ -20,21 +20,59 @@ export interface TracerEvent {
...
+  abstract evaluateEvent(args: {
+    event: TracerEvent;
+  }): Evaluation | Promise<Evaluation>;
+}
 
-  maxConcurrency: number = 10;
+/**
+ * An ABC for users that are implementing an evaluator that will be run against both test cases and production events.
+ */
+export abstract class BaseEvaluator<TestCaseType, OutputType>
```
Comment (nicolewhite) [inline src/testing/models.ts:68]: @dmorton2297 @adamnolte lmk what you think of this name

File: src/testing/run.ts [hunk 3]
```diff
@@ -244,11 +244,15 @@ export async function runTestSuite<
       throw new Error(`[${args.id}] No test cases provided.`);
     }
     args.evaluators.forEach((evaluator) => {
-      if (!(evaluator instanceof BaseTestEvaluator)) {
```
Comment (nicolewhite) [inline src/testing/run.ts:247]: we could have left this if BaseEvaluator extends BaseTestEvaluator but it's just a separate class

File: src/testing/models.ts (hunk 2 above)
Comment (adamnolte) [reply src/testing/models.ts:68]: We should move this out of testing and export it from both testing and tracer packages

Comment (nicolewhite) [reply src/testing/models.ts:68]: hmm we discussed this already and decided to keep everything in testing

Comment (adamnolte) [reply src/testing/models.ts:68]: It's pretty awkward to have to import the testing package into my main app when I'm just using the tracer

Comment (adamnolte) [reply src/testing/models.ts:68]: It's not a huge deal but it did take me a little bit to figure out how to import this in one of my demo apps where I was not using testing.

Comment (nicolewhite) [reply src/testing/models.ts:68]: where the discussion landed previously is that "testing" is relevant in both online and offline scenarios so i don'tt think it's weird to import from testing

Comment (nicolewhite) [reply src/testing/models.ts:68]: ok i'm open to changing we just need to keep these models together IMO, so maybe we can have a new thing called `evaluators` just for these types/models and we can import from there in both online and offline scenarios

Comment (adamnolte) [reply src/testing/models.ts:68]: Ya I think having an evaluators package makes sense.

Comment (nicolewhite) [reply src/testing/models.ts:68]: what do you think of the name though? wasn't sure if i should try to make it more clear that it's both in the name

Comment (adamnolte) [reply src/testing/models.ts:68]: I think the name makes sense

File: src/testing/models.ts [hunk 4]
```diff
@@ -20,21 +20,59 @@ export interface TracerEvent {
   properties: ArbitraryProperties;
 }
 
-export abstract class BaseTestEvaluator<TestCaseType, OutputType> {
+interface TestEvaluator<TestCaseType, OutputType> {
```
Comment (adamnolte) [inline src/testing/models.ts:23]: Thoughts on TestCaseEvaluator?

Comment (nicolewhite) [reply src/testing/models.ts:23]: sure but just FYI this is an internal type, not something an end user would use

Comment (nicolewhite) [reply src/testing/models.ts:23]: did you mean the evaluator? renaming `BaseTestEvaluator` to `BaseTestCaseEvaluator`?

Comment (adamnolte) [reply src/testing/models.ts:23]: ya i meant also updating `BaseTestEvaluator` to `BaseTestCaseEvaluator`

Comment (nicolewhite) [reply src/testing/models.ts:23]: i guess we could but the most accurate name is probably `TestCaseAndItsOutputEvaluator` :p so "test" seemed to capture all of that. we'd have to leave the old name as well for backwards compatibility

**Review by adamnolte — APPROVED** [review]
The export stuff can be a different task

File: src/testing/models.ts (hunk 4 above)
Comment (nicolewhite) [reply src/testing/models.ts:23]: i'm ok w/ renaming, i think at least `BaseTestCaseEvaluator` is more accurate than `BaseTestEvaluator`. can tackle along with the move to `evaluators` module?

Comment (adamnolte) [reply src/testing/models.ts:23]: Yep that sounds good.

**Pull request merged**